### PR TITLE
fix #12735 on osx, call dsymutil for debug builds

### LINKED ISSUE #12735: osx: dsymutil needs to be called for debug builds to keep debug info
on osx, `dsymutil` needs to be called on the resulting executable when user requests debug builds (eg `-g` or `--debugger:native`) otherwise DWARF debug info is gone; see rationale here [1]

Note that the DWARF debug info is needed for:
* improved debugging experience (symbolication) under a debugger
* https://github.com/nim-lang/Nim/issues/12702 (stacktraces without runtime overhead)
* symbolication of stacktraces pointing to c++ code (which could be emitted, from a loaded library etc)

tools like `llvm-symbolizer -obj=/pathto/prog` can do proper symbolication (address => file/line) if `/pathto/prog.dSYM` is alongside the binary, as would be the case if `dsymutil /pathto/prog` is called:

On OSX, when using non-separate compilation, `-g` generates a `foo.dSYM` dir containing DWARF debug info, as can be seen using -v:

```
# main.c is any c program, eg a hello world
clang -v -g -o /tmp/z12 main.c
# the last step shows:
dsymutil -o /tmp/z12.dSYM /tmp/z12
```

However, when using separate compilation (as is the case for nim), the last step `dsymutil` needs to be called explicitly by the build logic.

We can detect whether user requested debug build by `optCDebug in conf.globalOptions`. If so, we should call dsymutil on executable 

### proposal
```nim
void afterBuild(conf: ConfigRef):
 when defined(osx):
  # remove directory exe / ".dSYM" if exists
  if optCDebug in conf.globalOptions and (isDefined(c) or isDefined(cpp) or isDefined(objc)):
    # user requested a debug build, and it's not js or weird target
    let exe = conf.outDir / conf.outFile
    let cmd = "dsymutil " & quoteShell(exe)
    if execShellCmd(cmd) != 0:
      echo "dsymutil cmd failed: " & cmd # send a warning or error?
```

## note
the separation of debug info from the binary allows several use cases, eg shipping to customers a build stripped from debug info, while still allowing symbolication of stacktraces / core dumps given a (say, non-shipped) .dSYM directory that corresponds to the released binary 

[1] [Apple's "Lazy" DWARF Scheme - Dwarf Wiki](http://wiki.dwarfstd.org/index.php?title=Apple%27s_%22Lazy%22_DWARF_Scheme)

### PR DESCRIPTION
**Pull request by timotheecour**
/cc @stefantalpalaru it should help with https://github.com/status-im/nim-libbacktrace and https://github.com/nim-lang/Nim/pull/12922
/cc @Araq 

* fixes https://github.com/nim-lang/Nim/issues/12735

### CODE REVIEW

**Comment by timotheecour** [comment]
failure unrelated, caused by https://github.com/nim-lang/Nim/issues/12919

**Comment by Araq** [comment]
Beyond my knowledge. What does it do? Why is it required?

**Comment by stefantalpalaru** [comment]
> What does it do?

`dsymutil` collects those debugging symbols in object files (pointed at by a slim debugging section in the binary passed as an argument) and places them in a "<binary_name>.dSYM" directory next to the binary.

Why not just have the linker be able to add those debugging symbols in the linking step and rely on a different tool and a different step for that? Ask Apple.

> Why is it required?

So Nim's `--debugger:native` option starts working on macOS.

**Comment by timotheecour** [comment]
not needed for this PR, but more info on apple's dSYM design is found here: http://wiki.dwarfstd.org/index.php?title=Apple%27s_%22Lazy%22_DWARF_Scheme ; one aspect of separation of debug info into a separate directory is it allows shipping to customers a slim binary (with no debug info), and to symbolicate customer provided stacktraces given a dSYM that can be kept separately; tools like lldb etc ensure the correct dSYM is used for a given binary using a hash  to avoid wrong symbolications

File: compiler/extccomp.nim [hunk 1]
```diff
@@ -1077,6 +1090,9 @@ proc writeJsonBuildInstructions*(conf: ConfigRef) =
 
     lit "\L}\L"
     close(f)
+    # we really should use json apis (eg %* and %) instead of all those `lit`,
+    # no need for premature optimization
+    writeFile $jsonFile, ($jsonFile).parseFile.pretty
```
Comment (Araq) [inline compiler/extccomp.nim]: Remove this please, it has nothing to do with your patch and I don't like it.

Comment (timotheecour) [reply compiler/extccomp.nim]: done

**Comment by timotheecour** [comment]
PTAL

**Pull request merged**